feat: display entire element chain

thwilson3 opened the pull request:
## Problem

Continuing work from #16126 which aims to resolve #14347 

I'll still need to fix the tests but I've tried to replicate the desired results requested for #16126 

## Changes
_I apologize for the slight stutter in the video, my computer is struggling_

https://github.com/PostHog/posthog/assets/107593248/85d3e42d-5434-4c7f-95e7-5ddfa41a71d5

https://github.com/PostHog/posthog/assets/107593248/5191f434-1538-4f86-9f83-70402e0d6931

## How did you test this code?

Manual Testing:

- Expanded multiple events to test various lengths of event captures.
- Tested Toolbar view when editing an action
- Clicked the new "Show x parent elements (x elements hidden)" row to verify the element text was captured correctly.
- Verified functionality between every commit.
- Double checked all of the pain points mentioned in #16126 to ensure those bugs were addressed.

thwilson3 committed:
First pass at viewing additional elements in the element chain view

thwilson3 committed:
Refactored:

-Move snippet inside 'element.length' check
-Update styling to closer match discussed preference
-Add dynamic text to pluralize 'element' if more than one element

thwilson3 committed:
Changed verbiage to match existing PR

thwilson3 committed:
Adjusted text format for 'Show x hidden elements' to accurately reflect white space in event description under event explorer

thwilson3 committed:
Added Fade animation to hidden elements when shown

thwilson3 committed:
Fixed incorrect ternary for hidden elements

thwilson3 committed:
Changed verbiage to reflect agreed upon changes

thwilson3 committed:
Added fade animation to closing tags

thwilson3 committed:
Reformatted button text so it would display correctly in the event column of the event explorer tab

thwilson3 committed:
Remove element chain limit

thwilson3 committed:
Add data-attr tag for accurate analytics

thwilson3 committed:
Reformat text for code clarity

thwilson3 committed:
Bug fix in toolbar view for editing an action

nizarmah commented on frontend/src/lib/components/HTMLElementsDisplay/HTMLElementsDisplay.tsx:
```diff
@@ -114,16 +144,29 @@ export function HTMLElementsDisplay({
+                                } (${elementsToShowDepth} hidden)`}
+                            </pre>
+                        ) : null}
+                        <Fade visible={true} style={{ position: 'static' }}>
+                            <Tags
+                                elements={parsedElements}
+                                highlight={highlight}
+                                editable={editable}
+                                parsedCSSSelectors={parsedSelectors}
+                                onChange={(index, s) => setParsedSelectors({ ...parsedSelectors, [index]: s })}
+                            />
+                            <CloseAllTags elements={parsedElements} />
+                        </Fade>
```
Adding a Fade tag here isn't necessary. You already have it under `Tags` and `CloseAllTags`.

thwilson3 replied:
Resolved in ad73878

nizarmah commented on frontend/src/toolbar/actions/actionsTabLogic.tsx:
I believe there's a bug with the `attr_class` in the `toElementsChain` which can be addressed here

https://github.com/PostHog/posthog/blob/89d4dbd7da765a0881fcb9e0e124715451113954/frontend/src/toolbar/actions/actionsTabLogic.tsx#L32

thwilson3 replied:
Resolved in 2a44c29

nizarmah commented on frontend/src/lib/components/HTMLElementsDisplay/htmlElementsDisplayLogic.ts:
```diff
@@ -57,6 +57,18 @@ export const htmlElementsDisplayLogic = kea<htmlElementsDisplayLogicType>([
         ],
+        parsedElements: [
+            (s) => [s.elements, s.visibleElements],
+            (elements: ElementType[], visibleElements: number) => {
```
No need for types here because the types will be derived from the `(s) => [...]`.

thwilson3 replied:
Resolved in 6dbb4dc

nizarmah commented on frontend/src/lib/components/HTMLElementsDisplay/HTMLElementsDisplay.tsx:
```diff
@@ -93,8 +115,16 @@ export function HTMLElementsDisplay({
+    const { setParsedSelectors, showAdditionalElements } = useActions(logic)
 
     return (
         <div className="flex flex-col gap-1">
```
Below this line, we're still using `!!elements.length`, but I believe we should be using `!!parsedElements.length` and remove `elements` from the values used in this component.

thwilson3 replied:
Addressed in c7085c2

nizarmah commented on frontend/src/lib/components/HTMLElementsDisplay/HTMLElementsDisplay.tsx:
```diff
@@ -93,8 +115,16 @@ export function HTMLElementsDisplay({
+        chosenSelector,
+        chosenSelectorMatchCount,
+        messageStatus,
+        elements,
```
This is a follow-up comment

```suggestion
```

thwilson3 replied:
Addressed in c7085c2

nizarmah reviewed (changes requested):
Hey Tommy! Nice work on this 😃
 
Left a couple of comments.

thwilson3 committed:
Fix failing test

thwilson3 committed:
Update use of elements to parsedElements

pauldambra reviewed (commented):
Agree this looks like a good re-implementation of the previous PR

Let's get the other comments addressed and then we can see what is failing in the tests and understand why 🚀

thwilson3 committed:
Fix bug in attr_class

thwilson3 committed:
Fix bug where button text displayed incorrectly

thwilson3 committed:
Remove unnecessary Fade element

thwilson3 committed:
Remove unnecessary types, reformatted showAdditionalElements action to match Kea documentation

thwilson3 commented on frontend/src/lib/components/HTMLElementsDisplay/htmlElementsDisplayLogic.ts:
```diff
@@ -30,6 +30,7 @@ export const htmlElementsDisplayLogic = kea<htmlElementsDisplayLogicType>([
     actions({
         setParsedSelectors: (selectors: Record<number, ParsedCSSSelector>) => ({ selectors }),
         setElements: (providedElements: ElementType[]) => ({ providedElements }),
+        showAdditionalElements: () => ({}),
```
Resolved in 6dbb4dc

thwilson3 commented:
Thank you for the review @pauldambra! I appreciate your patience with me on this. I've implemented the requested changes and fixed the unit test that was failing. I believe the snapshot test is the last one failing. 

I'm currently troubleshooting a segmentation fault when trying to update the visual regressions, I'll hopefully have that figured out soon.

_**I also have a silly request, I've registered for Hacktoberfest and was hoping for this contribution to count toward it. Since it's opt-in for maintainers this year, the PR just needs to have the**_ `hacktoberfest-accepted` _**label in order to be eligible. I would really appreciate your help with that, thank you so much!**_

thwilson3 committed:
Merge branch 'master' of https://github.com/PostHog/posthog into display-entire-element-chain

nizarmah reviewed (approved):
Thanks for addressing my comments! 

I'm approving because I don't have anything to add.
I'll keep an eye on the reviews others leave though.

If you need any help, don't hesitate to hit me up!

pauldambra commented:
OK, just one test failure now... we need to update the visual regression snapshots. What happens if you run `pnpm test` @thwilson3 ?

thwilson3 commented:
Hey @pauldambra! So when I run `pnpm test` all of the frontend tests run normally. 

When digging through the logs on here for the failed test, I found that the failing test was for `frontend/src/lib/components/HTMLElementsDisplay/HTMLElementsDisplay.stories.tsx`, which I thought made sense so I read how to update the tests on the PostHog site. I started up storybook no problem but when I try to update the snapshot running `pnpm test:visual-regression:stories frontend/src/lib/components/HTMLElementsDisplay/HTMLElementsDisplay.stories.tsx` I get a segmentation fault error. 
<img width="831" alt="Screenshot 2023-10-12 at 6 26 55 PM" src="https://github.com/PostHog/posthog/assets/107593248/072ec091-57bf-426e-a21d-aa2a08c7cce9">

My guess is that my computer is struggling with the memory demands. For additional context, I have a 2020 M1 Macbook Air with 8GB of ram. I have everything in Docker running (I've bumped the allocated memory for Docker up to 7GB but still hit the issue), all of the Frontend/Backend/Workers running from using `./bin/start` and then I run storybook shortly before trying to update the snapshot. 

I also saw the `WARN` message about the Node version, so I updated and tried to point to using that version but I'm not sure if the command is configured to run a certain way or if I'm doing something wrong on my end.

_thank you for adding the label by the way, I really appreciate that!_

pauldambra commented:
@thwilson3 I'll merge this and then fix up the snapshots separately rather than try and overheat your laptop to run them 🤣 

Thanks for the contribution @thwilson3 @nizarmah 🚀

thwilson3 commented:
Awesome, thank you so much for the help on that!! Happy to be able to contribute!!

nizarmah commented:
My pleasure! Kudos to @thwilson3 for picking it up 🙌🏼 And as tommy already mentioned, huge thanks on all the help ❤️
